src/charm.py

```diff
@@ -1,103 +1,202 @@
+        """
+        for param in config_list:
+            if self.config[param].strip() == "":
+                raise ValueError(f"Invalid config: {param} value missing")
+
+    def _update(self, event):
+        """Update the Temporal worker configuration and replan its execution.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        try:
+            self._validate()
+        except ValueError as err:
+            self.unit.status = BlockedStatus(str(err))
+            return
+
+        container = self.unit.get_container(self.name)
+        if not container.can_connect():
+            event.defer()
+            return
+
+        # ensure the container is set up
+        _setup_container(container)
+
+        logger.info("Configuring Temporal worker")
+
+        pebble_layer = {
+            "summary": "temporal worker layer",
             "services": {
-                "httpbin": {
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
+                self.name: {
+                    "summary": "temporal worker",
+                    "command": f"python worker.py '{json.dumps(dict(self.config))}'",
                     "startup": "enabled",
-                    "environment": {
-                        "GUNICORN_CMD_ARGS": f"--log-level {self.model.config['log-level']}"
-                    },
+                    "override": "replace",
                 }
             },
         }
 
+        container.add_layer(self.name, pebble_layer, combine=True)
+        container.replan()
+
+        self.unit.status = ActiveStatus()
+
+
+def _setup_container(container: Container):
+    """Copy worker file to the container and install dependencies.
+
+    Args:
+        container: Container unit on which to perform action.
+    """
+    resources_path = Path(__file__).parent / "resources"
+    _push_container_file(container, resources_path, "/worker.py", resources_path / "worker.py")
+    _push_container_file(
+        container,
+        resources_path,
+        "/temporal_client/__init__.py",
+        resources_path / "temporal_client/__init__.py",
+    )
+    _push_container_file(
+        container,
+        resources_path,
+        "/temporal_client/workflows.py",
+        resources_path / "temporal_client/workflows.py",
+    )
+    _push_container_file(
+        container,
+        resources_path,
+        "/temporal_client/activities.py",
+        resources_path / "temporal_client/activities.py",
+    )
+
+    # Install worker dependencies
+    worker_dependencies_path = resources_path / "worker-dependencies.txt"
+    with open(worker_dependencies_path, "r") as dependencies_file:
+        dependencies = dependencies_file.read().split("\n")
+        logger.info(f"installing worker dependencies {dependencies}...")
+        container.exec(["pip", "install", *dependencies]).wait()
```

Comment:
I think this is fine. Just wondering if you considered `pip install -r {dependencies}`. Also, do we need validation? What happens if someone puts a line like `; rm -rf /` in the dependency file?

Reply:
For now, these are the necessary Temporal dependencies bundled with the charm. When we add an action for allowing a user to install dependencies on the workload container, I will make sure to add validation.

Reply:
+1 on validation

Reply:
Good enough for me.